torch/jit.py

```diff
@@ -1,26 +1,375 @@
+
+    Example: Trace as function decorator.
+
+        >>> @jit.trace
+        >>> def f(x);
+        >>>     return x * 2
+        >>> trace, out = f(Variable(torch.randn(1)))
+
+    Example: Trace as higher order function. (Notice that trace is a *curried*
+    function; you first apply it with the function/model to trace, and then
+    apply the result with the arguments.)
+
+        >>> traced_model = jit.trace(nn.LSTMCell())
+        >>> trace, out = traced_model(input, hidden)
+
+    Example: Trace the backwards pass as function decorator.
+
+        >>> @jit.trace(nderivs=1)
+        >>> def f(x);
+        >>>     return x * 2
+        >>> trace, out = f(Variable(torch.randn(1)))
+        >>> out.sum().backward()
+        >>> print(trace)
+
+    Example: Trace the backwards pass as higher order function.
+
+        >>> traced_model = jit.trace(nn.LSTMCell(), nderivs=1)
+        >>> trace, out = traced_model(input, hidden)
+        >>> out.sum().backward()
+        >>> print(trace)
+
+    Example: Trace a function with extra parameters. (If you trace
+    a Module, parameters are automatically computed via `state_dict`).
+
+        >>> lstm = nn.LSTMCell(10, 20)
+        >>> @jit.trace(params=lstm.parameters())
+        >>> def f(a, b):
+        >>>     return lstm(a, b)
+    """
+    # TODO: handle decorating a class (not a callable)
+    def _trace(inner):
+        return TracedModule(inner, nderivs=nderivs, params=params)
+    if callable(arg):
+        return _trace(arg)
+    else:
+        return _trace
+
+
+# TODO: Formulating it this way means that state_dict of the traced thing
+# looks different, etc (because there's extra nesting).  Figure out a
+# way to avoid this
+class TracedModule(Module):
+    def __init__(self, inner, params=tuple(), nderivs=0):
+        super(TracedModule, self).__init__()
+        # inner may be a Module, or it may be an arbitrary callable
+        self.inner = inner
+        self.params = ParameterList(list(params))
+        self.nderivs = nderivs
+
+    def forward(self, *args):
+        # TODO: Possible optimization: use the unflattened
+        # output so we don't unflatten it when we get out
+        # NB: Not a method because trace_func_raw can't deal
+        # with methods
+        @raw_trace(nderivs=self.nderivs)
+        def traced_inner(in_vars, in_struct):
+            # NB: Uncommenting this line should be equivalent
+            # args, _params = _unflatten(in_vars, in_struct)
+            return _flatten(self.inner(*args))
+
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        trace, (out_vars, out_struct) = traced_inner(in_vars, in_struct)
+        out, extra = _unflatten(out_vars, out_struct)
+        assert len(extra) == 0
+        return trace, out
+
+
+# Functional version that assumes that all parameters are explicitly
+# specified
+def raw_trace(nderivs=0):
+    def _raw_trace(f):
```

Comment:
Can you document what's expected of `f`? It doesn't seem to be something that should be user-facing so prepend an underscore